components/board/genericlinux/data.go

```diff
@@ -92,75 +75,44 @@ func readIntFile(filePath string) (int, error) {
 	return int(resultInt64), err
 }
 
-func getGpioChipDefs(pinDefs []PinDefinition) (map[string]gpioChipData, error) {
-	gpioChipsInfo := map[string]gpioChipData{}
-	sysfsPrefixes := []string{"/sys/devices/", "/sys/devices/platform/", "/sys/devices/platform/bus@100000/"}
+func getGpioChipDefs(pinDefs []PinDefinition) (map[int]gpioChipData, error) {
+	gpioChipsInfo := map[int]gpioChipData{}
 
-	// Get a set of all the chip names with duplicates removed. Go doesn't have native set objects,
-	// so we use a map whose values are ignored.
-	gpioChipNames := make(map[string]struct{}, len(pinDefs))
-	for _, pinDef := range pinDefs {
-		if pinDef.GPIOChipSysFSDir == "" {
-			continue
-		}
-		gpioChipNames[pinDef.GPIOChipSysFSDir] = struct{}{}
-	}
-
-	for gpioChipName := range gpioChipNames {
-		var gpioChipDir string
-		for _, prefix := range sysfsPrefixes {
-			d := prefix + gpioChipName
-			fileInfo, err := os.Stat(d)
-			if err != nil {
-				continue
-			}
-			if fileInfo.IsDir() {
-				gpioChipDir = d
-				break
-			}
-		}
-		if gpioChipDir == "" {
-			return nil, errors.Errorf("cannot find GPIO chip %q", gpioChipName)
-		}
-		files, err := os.ReadDir(gpioChipDir)
+	allDevices := gpio.ChipDevices()
+	gpioChipNgpios := make(map[int]string, len(allDevices)) // maps chipNgpio -> string gpiochip#
+	for _, dev := range allDevices {
+		chip, err := gpio.OpenChip(dev)
 		if err != nil {
 			return nil, err
 		}
-		var chipFileName string
-		for _, file := range files {
-			if !strings.HasPrefix(file.Name(), "gpiochip") {
-				continue
-			}
-			chipFileName = file.Name()
-			break
-		}
 
-		gpioChipGPIODir := gpioChipDir + "/gpio"
-		files, err = os.ReadDir(gpioChipGPIODir)
+		chipInfo, err := chip.Info()
 		if err != nil {
 			return nil, err
 		}
-		for _, file := range files {
-			if !strings.HasPrefix(file.Name(), "gpiochip") {
-				continue
-			}
 
-			base, err := readIntFile(filepath.Join(gpioChipGPIODir, file.Name(), "base"))
-			if err != nil {
-				return nil, err
-			}
+		gpioChipNgpios[int(chipInfo.NumLines)] = chipInfo.Name
```

Comment:
If we ever encounter a board with 2 chips with the same ngpio value, we're gonna have a bad time. Log an error if that happens, so we notice the issue:
```go
if _, ok = gpioChipNgpios[int(chipInfo.NumLines)]; ok {
    logger.Global().Errorf("Board has multiple GPIO chips with the same ngpio value, %d!", chipInfo.NumLines)
}
gpioChipNgpios[int(...)] = chipInfo.Name
```